envoy/npds.proto

```diff
@@ -0,0 +1,127 @@
+  }
+
+  rpc FetchNetworkPolicies(envoy.api.v2.DiscoveryRequest) returns (envoy.api.v2.DiscoveryResponse) {
+    option (google.api.http) = {
+      post: "/v2/discovery:network_policies"
+      body: "*"
+    };
+  }
+}
+
+// A network policy to be enforced on the network flows to/from associated hosts.
+message NetworkPolicy {
+  // The unique identifier of the network policy.
+  // Required.
+  uint64 policy = 1;
+
+  // The identifier of a filter that enforces a subset of the policy.
+  // If specified, only the subset of the policy that is to be enforced by this
+  // filter is contained in `ingress` and `egress`.
+  //
+  // Optional. If omitted, the full policy is contained in `ingress` and
+  // `egress`.
+  string filter = 2;
+
+  // The part of the policy to be enforced at ingress.
+  // Optional.
+  DirectionNetworkPolicy ingress = 3;
+
+  // The part of the policy to be enforced at egress.
+  // Optional.
+  DirectionNetworkPolicy egress = 4;
+}
+
+// A network policy to be enforced on flows either to or from hosts.
+message DirectionNetworkPolicy {
+  // The set of per-port network policies, one per destination L4 port.
+  // Every PortNetworkPolicy element in this set has a unique toPort.
+  // Required and may not be empty.
+  repeated PortNetworkPolicy per_port_policies = 1 [(validate.rules).repeated .min_items = 1];
+}
+
+// A network policy to be enforced on flows to a specific destination L4 port.
+message PortNetworkPolicy {
+  // The flows' destination L4 port number, as an unsigned 16-bit integer.
+  // Required.
+  uint32 port = 1 [(validate.rules).uint32.lte = 65535];
+
+  // The flows' L4 transport protocol.
+  // Required.
+  envoy.api.v2.SocketAddress.Protocol protocol = 2;
+  // TODO: Fix protoc-gen-validate. Adding any validate.rules to protocol above results in generating broken C++ code.
+
+  // The network policy rules to be enforced on the flows to the port.
+  // A flow is allowed if any rule matches it. If no rule matches the flow, it is denied.
+  // Required and may not be empty.
+  repeated PortNetworkPolicyRule rules = 3 [(validate.rules).repeated .min_items = 1];
+}
+
+// A network policy rule, as a conjunction of predicates on L3-L7 flows.
+// If all the predicates of a rule match a flow, the flow is allowed.
+message PortNetworkPolicyRule {
+  // The set of identifiers of policies of remote hosts that are allowed by this rule.
+  // This predicate matches a flow if the identifier of the policy applied on the flow's remote
+  // host is contained in this set. Required and may not be empty.
+  repeated uint64 remote_policies = 1 [(validate.rules).repeated = {min_items: 1, unique: true}];
+
+  oneof l7_rules {
+    // The set of HTTP network policy rules.
+    // An HTTP request is allowed if any rule in this set matches it.
+    // Optional.
+    HttpNetworkPolicyRules http_rules = 100;
+ }
+}
+
+// A set of network policy rules enforced on HTTP requests.
+message HttpNetworkPolicyRules {
```

Comment:
Maybe this should be named as HttpPolicyRules instead? Envoy has quite a strong separation between e.g., network filters and http filters, network and http namespaces, etc.

Reply:
"Policy" is heavily overloaded. I wanted to always use the term "network policy" to avoid any ambiguity.

Reply:
No strong opinion here, just observing that Envoy keeps "Network" (== L3/L4) separate from "HTTP" (=L7), in naming as well as in code organization.